### PR TITLE
Use navigator.share on mobile devices

### LINKED ISSUE #9962: Native sharing not working
**Describe the bug**

When I tap the ellipsis, I don't see a native share dialog.

**To Reproduce**

1. Open an article
2. Tap the ellipsis

**Expected behavior**

I expected to see a native share dialog, as depicted in #1524

**Screenshots**

**Desktop (please complete the following information):**

N/A

**Smartphone (please complete the following information):**

- Device: Google Pixel 4
- OS, version: Android 10
- Browser, version: Chrome 84.0.4147.125

**Additional context**

Not familiar with this codebase, but it looks like this area may have been touched in #6815

### PR DESCRIPTION
**Pull request by fdoxyz**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Bug Fix
- [x] Optimization

## Description

We now have a tool for filtering UI based on the current runtime context so we can reliably separate Desktop vs Mobile UI to display some features we want on either one of them.

We had the [web-share-wrapper](https://www.npmjs.com/package/web-share-wrapper) documented for quite a while because we weren't able to have a bit more fine grain control over it, but this PR now makes use of it on mobile devices (where `navigator.share` API [is available]()).

## Related Tickets & Documents

Closes #9962

## QA Instructions, Screenshots, Recordings

1. Boot Forem locally
2. Make sure the actions on a Post continue to work the same as before in a Desktop Browser
3. Browse your local environment using a mobile device **if possible**
   - Instead of using anchor tag intents the menu should only display a single `Share` option
   - Tapping on this `Share` option should use the native share from the OS
   - Android and iOS should both work the same
   - Desktop browser DevTools "mobile simulator" won't work because the Forem instance will still detect the context as a "Desktop browser"

**Desktop view**

Share intents are always visible and _"Share Post via..."_ will be presented when `navigator.share` API is available

Screen Shot 2021-06-08 at 10 19 49

**Mobile view (native share within browser)**

IMG_635951DCFF42-1

### UI accessibility concerns?

None

## Added tests?

- [ ] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

- [x] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post

## [optional] What gif best describes this PR or how it makes you feel?

shrek donkey smile

### CODE REVIEW

File: app/assets/stylesheets/runtime.scss [hunk 1]
```diff
@@ -8,89 +8,67 @@
...
-.PWA-Windows,
-.PWA-Linux,
-.PWA-macOS,
-.Browser,
-.Browser-iOS,
-.Browser-Android,
-.Browser-macOS,
-.Browser-Windows,
-.Browser-Linux {
+.ForemWebView-only,
```
Comment (fdoxyz) [inline app/assets/stylesheets/runtime.scss:11]: This change was discussed here: 

https://github.com/forem/forem/pull/13662#discussion_r626903852

**Comment by fdoxyz** [comment]
Thank you for reaching out with this question/suggestion @aitchiss!

Sadly we still need that for now. Despite `navigator.share` being supported in Android browsers, the Android WebView implementation (within the native app) doesn't support it. `Runtime.isNativeAndroid('shareText')` checks to make sure we're on native android context **and** the shareText feature is supported, which will trigger the native share sheet manually (from Kotlin code).

The positive side of this is that this native share sheet allows for "copy to clipboard" of the URL and also sharing directly with other apps 🙂

I'm working on a E2E test for this feature (taking into account browser, mobile browser and native app support) and will be taking it out from Draft hopefully soon 🤞

**Comment by aitchiss** [comment]
Good to know! Thanks @fdoxyz 🙂

I was mostly wondering because I'm working on some dropdown a11y optimisations at the moment and had spotted this code along the way this week. Good to get my head around how it's all hanging together in the various places!

**Comment by thomasbnt** [comment]
Nice idea ! 🙌🏼

**Review by fdoxyz — COMMENTED** [review]
I attempted to add a Cypress test but was unsuccessful because I wasn't able to mock `navigator.share` API. I tried a few variations of these:

```js
cy.visit(response.body.current_state_path, function() {
   onBeforeLoad: (win) => {
      Object.defineProperty(win.navigator, 'share', () => { return true });
   }
});

cy.visit(response.body.current_state_path, function() {
   onBeforeLoad: (win) => {
      Object.defineProperty(win.navigator, 'share', { value: true });
   }
});
```

**Comment by nickytonline** [comment]
> I attempted to add a Cypress test but was unsuccessful because I wasn't able to mock `navigator.share` API. I tried a few variations of these:
> 
> ```js
> cy.visit(response.body.current_state_path, function() {
>    onBeforeLoad: (win) => {
>       Object.defineProperty(win.navigator, 'share', () => { return true });
>    }
> });
> 
> cy.visit(response.body.current_state_path, function() {
>    onBeforeLoad: (win) => {
>       Object.defineProperty(win.navigator, 'share', { value: true });
>    }
> });
> ```

I'll take a peek at this @fdoxyz.

**Comment by nickytonline** [comment]
> > I attempted to add a Cypress test but was unsuccessful because I wasn't able to mock `navigator.share` API. I tried a few variations of these:
> > ```js
> > cy.visit(response.body.current_state_path, function() {
> >    onBeforeLoad: (win) => {
> >       Object.defineProperty(win.navigator, 'share', () => { return true });
> >    }
> > });
> > 
> > cy.visit(response.body.current_state_path, function() {
> >    onBeforeLoad: (win) => {
> >       Object.defineProperty(win.navigator, 'share', { value: true });
> >    }
> > });
> > ```
> 
> I'll take a peek at this @fdoxyz.

@fdoxyz, it works if you use your second attempt.

```javascript
Object.defineProperty(win.navigator, 'share', {
	value: true,
});
```

image

**Comment by fdoxyz** [comment]
Whoa, I think I have a problem with my local Cypress execution... I went ahead and pushed a commit that added the assert to the Cypress test and it passed in Travis 🤯 

It might be related with how I generally run my local environment with ngrok, but I guess it's time for me to nuke my local env (after quite a while of not doing it) to be sure things like this work locally for me 😞 

Thank you for double checking this @nickytonline!

**Comment by aitchiss** [comment]
I gave this a quick test run on desktop (Safari/Chrome/Firefox) and all worked great 👍 

I also tried on Chrome web browser on my mobile with Android 11 but:
- there was no "Share via..." option
- the "copy to clipboard" functionality didn't work


https://user-images.githubusercontent.com/20773163/121677198-7a017180-caad-11eb-9a48-53fecb414441.mp4

File: app/javascript/packs/articlePage.jsx [hunk 2]
```diff
@@ -47,7 +48,18 @@ if (shareDropdownButton.dataset.initialized !== 'true') {
     document
       .querySelectorAll('#article-show-more-dropdown [href]')
       .forEach((link) => link.addEventListener('click', closeDropdown));
+
+    // Add Ahoy stats for dropdown share options
+    const container = document.getElementsByClassName(
+      'crayons-article-actions__inner',
+    )[0];
```
Comment (aitchiss) [inline app/javascript/packs/articlePage.jsx:55]: suggestion (not blocking) - you could use `querySelector` which will stop looking after it finds the first one. In practice it doesn't matter much though, as there will only be one on the page anyway 😄 

```suggestion
    const container = document.querySelector(
      '.crayons-article-actions__inner',
    );
```

File: app/javascript/packs/articlePage.jsx [hunk 3]
```diff
@@ -47,7 +48,18 @@ if (shareDropdownButton.dataset.initialized !== 'true') {
...
+    container?.querySelectorAll('a.dropdown-link-row').forEach((link) => {
+      link.addEventListener('click', (event) => {
+        ahoy.track('Post Dropdown', { option: event.target.text.trim() });
+      });
+    });
```
Comment (aitchiss) [inline app/javascript/packs/articlePage.jsx:60]: We already grab and loop through all of these links on line 48, so I think it would be more efficient to add these event listeners in that same loop, e.g:

```js
 // We want to close the dropdown on link select (since they open in a new tab)
 // We also have Ahoy stats for dropdown share options
    document
      .querySelectorAll('#article-show-more-dropdown [href]')
      .forEach((link) => {
        link.addEventListener('click', closeDropdown);
        ahoy.track('Post Dropdown', { option: event.target.text.trim() });
      });
```

File: app/views/articles/_actions.html.erb [hunk 4]
```diff
@@ -76,10 +77,14 @@
             href="https://www.facebook.com/sharer.php?u=<%= article_url(@article) %>">
             Share to Facebook
           </a>
+        </div>
+
+        <web-share-wrapper shareurl="<%= article_url(@article) %>" sharetitle="<%= @article.title %>" sharetext="<%= @article.description %>" template="web-share-button">
         </web-share-wrapper>
         <template id="web-share-button">
-          <a href="#" class="dropdown-link-row crayons-link crayons-link--block">Share Post</a>
+          <a href="#" class="dropdown-link-row crayons-link crayons-link--block">Share Post via...</a>
```
Comment (aitchiss) [inline app/views/articles/_actions.html.erb:85]: From an accessibility POV this should really be a button, as it executes an action on the current page, rather than linking anywhere.

In saying that, if we make it a button we'd need to consider the design a bit more, as it would/should look different than the links.

I guess what I'm saying is I would _like_ to say this is a blocking issue, but I appreciate it might not be super-straightforward to just change 😅

File: cypress/integration/articleFlows/postSidebarActions.spec.js [hunk 5]
```diff
@@ -37,6 +43,8 @@ describe('Post sidebar actions', () => {
     cy.findByRole('link', { name: /^Share to Reddit$/i });
     cy.findByRole('link', { name: /^Share to Hacker News$/i });
     cy.findByRole('link', { name: /^Share to Facebook$/i });
+    // The navigator.share API should be available making this option visible
+    cy.findByRole('link', { name: /^Share Post via...$/i });
```
Comment (aitchiss) [inline cypress/integration/articleFlows/postSidebarActions.spec.js:47]: We should probably have a separate test that checks it's not available when the stub isn't added - what do you think?

File: cypress/integration/articleFlows/postSidebarActions.spec.js [hunk 6]
```diff
@@ -1,4 +1,10 @@
 describe('Post sidebar actions', () => {
+  const runtimeStub = {
+    onBeforeLoad: (win) => {
+      Object.defineProperty(win.navigator, 'share', { value: true });
+    },
+  };
+
```
Comment (aitchiss) [inline cypress/integration/articleFlows/postSidebarActions.spec.js:7]: it looks so simple but I know it's not 😂  I'm glad we have this in place now and can replicate where needed! 🎉

File: jest.config.js [hunk 7]
```diff
@@ -28,7 +28,7 @@ module.exports = {
       statements: 42,
       branches: 38,
       functions: 41,
-      lines: 43,
+      lines: 42,
```
[the remark this answers is not in the document]
Comment (aitchiss) [reply jest.config.js:31]: Personally, I think it is fine to reduce this number for now. I know @nickytonline has had some discussions about getting Cypress coverage taken into account for our thresholds, and for now I'd rather we didn't add extra tests we don't need, just to meet a coverage level. As long as we feel the feature is tested in Cypress, I think we're good.

If we think it's needed, we can also test the ahoy event was sent by using Cypress - https://docs.cypress.io/guides/guides/network-requests#Testing-Strategies (I'm not sure we test the sending of analytics elsewhere so I wouldn't say it's essential)

File: app/javascript/packs/articlePage.jsx [hunk 8]
```diff
@@ -47,17 +47,13 @@ if (shareDropdownButton.dataset.initialized !== 'true') {
...
-    )[0];
-    container?.querySelectorAll('a.dropdown-link-row').forEach((link) => {
-      link.addEventListener('click', (event) => {
-        ahoy.track('Post Dropdown', { option: event.target.text.trim() });
+      .forEach((link) => {
+        link.addEventListener('click', closeDropdown);
+        // Temporary Ahoy Stats for usage reports
+        link.addEventListener('click', (event) => {
+          ahoy.track('Post Dropdown', { option: event.target.text.trim() });
+        });
```
Comment (nickytonline) [inline app/javascript/packs/articlePage.jsx:55]: Refactor (blocking): Two click event listeners is not necessary. Merge the body of each of these event listener callbacks into one callback.

```suggestion
        link.addEventListener('click', (event) => {
          closeDropdown(event)
          
          // Temporary Ahoy Stats for usage reports
          ahoy.track('Post Dropdown', { option: event.target.text.trim() });
        });
```

or add the ahoy tracking directly in `closeDropdown`.

**Review by nickytonline — CHANGES_REQUESTED** [review]

File: app/javascript/packs/articlePage.jsx (hunk 8 above)
Comment (fdoxyz) [reply app/javascript/packs/articlePage.jsx]: Thanks @nickytonline, this is indeed a much better way to do it!

**Comment by nickytonline** [comment]
> Personally, I think it is fine to reduce this number for now. I know @nickytonline has had some discussions about getting Cypress coverage taken into account for our thresholds, and for now I'd rather we didn't add extra tests we don't need, just to meet a coverage level. As long as we feel the feature is tested in Cypress, I think we're good.

Yeah, we've had some talks at the roundtables and I mentioned this to @djuber. I'll probably sync up with Dan for this. This threshold in jest made sense when we had no E2E tests, but now that we do, it doesn't make sense at all. We should fail on the aggregated coverage if it goes down or adjust it if we remove code.

TLDR 42 ftw!

**Comment by fdoxyz** [comment]
Okay, I think the feedback has been addressed and Travis is green again 😄 

> - there was no "Share via..." option
> - the "copy to clipboard" functionality didn't work

@aitchiss from your video I see a warning (in the URL bar) that makes me think you were browsing the ngrok URL with `http` and not `https`. Without `https` both `navigator.share` and `navigator.clipboard` aren't available (they require a secure context). Whenever you get a chance to try it again, could you explicitly add `https://` to the URL? 🙂 

Also, Chrome/Firefox on macOS don't support `navigator.share` but Safari does support it. On mobile the adoption is much broader.

**Comment by aitchiss** [comment]
> @aitchiss from your video I see a warning (in the URL bar) that makes me think you were browsing the ngrok URL with http and not https. Without https both navigator.share and navigator.clipboard aren't available (they require a secure context). Whenever you get a chance to try it again, could you explicitly add https:// to the URL? 🙂

Gah! Yes of course - woops! Double checked now on android and all working as expected 😄

File: cypress/integration/articleFlows/postShareOptions.spec.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+          published: true,
+        }).then((response) => {
+          articlePath = response.body.current_state_path;
+        });
+      });
+    });
+  });
+
+  it('should display "Share Post via..." when navigator.share is available', () => {
+    cy.visit(articlePath, shareAvailableStub);
```
Comment (aitchiss) [inline cypress/integration/articleFlows/postShareOptions.spec.js:29]: I love how neat & straightforward this is to use in each test variant 😍

**Review by aitchiss — APPROVED** [review]
This looks good to me!

Gave it another spin locally on both desktop and android mobile, and all works as outlined in the description above 🙌 

Like I mentioned before, the "Share post via..." should really be a button and not a link, but maybe that can be raised in a separate issue so we can think about what that would look like. In the meantime, this works grand ✨

**Review by nickytonline — APPROVED** [review]
Thanks for making the changes @fdoxyz! 🚢 🚢 🚢
